stream_flash minor fixes

de-nordic opened the pull request:
The batch of changes that include:
 1) Use flash_get_parameters to get erase value for stream_flash_buffered_write, where read of erased flash has been used previously;
 2) The ctx->buf_bytes check, within flash_sync, has been moved outside CONFIG_STREAM_FLASH_ERASE conditional code, as it is universal check whether there is any work to do;
 3) The update, within flash_sync,  of flash_stream_ctx bytes_written and buf_bytes is not done when verification callback fails;
 4) The stream_flash_erase_page will only update last_erased_page_start_offset if successful.

de-nordic committed:
storage/stream/flash: Don't update last erased page offset on failure

The stream_flash_erase_page would update stream_flash_ctx member
last_erased_page_start_offset, to page offset it attempted to erase,
even if such operation failed.
The commit changes this behaviour so that in case of failure the
last_erased_page_start_offset would still hold previously, successfully,
erase page offset.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/stream/flash: Fix flash_sync updating bytes_written on failure

The flash_sync function is able to call, if specified, write
verification callback to check whether data flash has been correctly
written to a flash.  Part of that procedure is to read date back of
the flash and call the verification callback on the buffer;
in case if the read would fail, the flash_sync would return,
with an error code, without updating stream_flash_ctx.

The same logic should be applied to failed verification, but, due to
missing return, the stream_flash_ctx gets updated with probably
incorrectly written bytes added to total bytes_written and buf_bytes,
representing number of bytes awaiting in buffer, being zeroed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/stream/flash: flash_sync: Check if there is anything to write

The commit moves block that checks if flash_stream_ctx.buf_bytes
is non-zero out of CONFIG_STREAM_FLASH_ERASE specific code,  as that is
universal check and is valid also for not CONFIG_STREAM_FLASH_ERASE
code.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/stream/flash: Use flash_get_parameters to get flash erase_value

The stream_flash_buffered_write used to read byte from flash, in front
of the current buffer pointer, to obtain value of errased flash;
the code has been replaced with obtaining the value from flash
parameters.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

hakonfam commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -42,6 +41,8 @@ int stream_flash_erase_page(struct stream_flash_ctx *ctx, off_t off)
 	if (rc != 0) {
 		LOG_ERR("Error %d while erasing page", rc);
+	} else {
+		ctx->last_erased_page_start_offset = page.start_offset;
```
Could we add a unit test case for this?

de-nordic replied:
Yes, we should. Sorry I have completely forgotten to write tests for this while stuck on on other tasks. I will add tests for the changes.

hakonfam commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -97,6 +99,7 @@ static int flash_sync(struct stream_flash_ctx *ctx)
 		rc = ctx->callback(ctx->buf, ctx->buf_bytes, write_addr);
 		if (rc != 0) {
 			LOG_ERR("callback failed: %d", rc);
+			return rc;
```
Could we add a unit test case for this?

hakonfam reviewed (approved):
LGTM, just wondering if we should have test cases for the fixes to avoid regression.

nvlsianpu reviewed (changes requested):
fix the last commit message title.

de-nordic committed:
storage/stream/flash: Failed write of reminder rewinds buffer offset

The stream_flash_buffered_write, when invoked to do flush write, will
attempt to write the tail bytes from the buffer, filling the required
minimal write block size with erase value bytes; after write it rewinds
the buffer offset, bytes_written, by number of the "filler bytes".
Doe to lack of return code processing from flash_sync call, two things
would happen to context in case of failure:
 1) the ctx->bytes_written would be rewind pass the value it had before
    function call as it gets decremented by "filler bytes" even if write
    failed;
 2) the ctx->buf_bytes offset would be accounting for added "filler
    bytes" which should not be counted as data in buffer.

Proper processing of return code has been added to remove effects
described above.

Unit tests have been expended to cover the scenarios.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
@hakonfam I am still working on the tests but in meantime I have added this to the stack: https://github.com/zephyrproject-rtos/zephyr/pull/32488/commits/ba6782601fc6ccde968b07add88b235f4702f3b6, can you look at that one?

hakonfam commented:
> @hakonfam I am still working on the tests but in meantime I have added this to the stack: [ba67826](https://github.com/zephyrproject-rtos/zephyr/commit/ba6782601fc6ccde968b07add88b235f4702f3b6), can you look at that one?

Good catch! Again a unit test verifying the fix would be great. What do you think about providing `fill_length` as an argument to `flash_sync`? Maybe even the whole handling of the filling of the tail? As we already see, the current split of this makes it prone for bugs where the state is not updated correctly.

de-nordic commented:
> Again a unit test verifying the fix would be great. 

I am working on them. Unfortunately this requires overriding flash API to make functions fail at specific moments.

> What do you think about providing `fill_length` as an argument to `flash_sync`? Maybe even the whole handling of the filling of the tail? As we already see, the current split of this makes it prone for bugs where the state is not updated correctly.

The buffer alignments should happen closer to the `flash_write`, so yes probably moving them to `flash_sync` would be good idea.
That would also reduce arithmetic on correcting the ctx after failed operations as that can also be done closer to the source of failure.

hakonfam commented:
> Would you be willing to do such change?

Yes, I will get this done.

de-nordic commented:
> 
> 
> > Would you be willing to do such change?
> 
> Yes, I will get this done.

I think we should have some rfc for the stream_flash changes.
While fixing the issues I have found that it is hard to recover from failed write attempts.
For example: when you attempt to write N lengths of buffer and buffer already has some data, in case of failure you have no way to actually find out at which byte, at which buffer the write failed; this means that when write of first buffer fails, you have probably lost data that you have no control over, because it has been placed in buffer with some previous calls to write.

nvlsianpu commented:
@de-nordic Thanks for `a bad weather` testcases.
